src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,581 @@
+                 * and also quote any left over (and invalid) % specifier from
+                 * the desktop file.
+                 */
+                raw = strreplace(c, "%%", "%");
+                if (!raw)
+                        return log_oom();
+                p = strreplace(raw, "%", "%%");
+                if (!p)
+                        return log_oom();
+                escaped = cescape(p);
+                if (!escaped)
+                        return log_oom();
+
+                if (asprintf(&quoted, "\"%s\"", escaped) < 0)
+                        return log_oom();
+
+                free(exec_split[n]);
+                exec_split[n++] = TAKE_PTR(quoted);
+        }
+        for (; exec_split[n]; n++)
+                exec_split[n] = mfree(exec_split[n]);
+
+        res = strv_join(exec_split, " ");
+        if (!res)
+                return log_oom();
+
+        *ret_exec_start = res;
+        return 0;
+}
+
+int xdg_autostart_service_generate_unit(
+                XdgAutostartService *service,
+                const char *dest) {
+
+        _cleanup_free_ char *path_escaped = NULL;
+        _cleanup_free_ char *exec_start = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        const char *unit;
+        int r;
+
+        assert(service);
+
+        /* Nothing to do for hidden services. */
+        if (service->hidden) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (service->systemd_skip) {
+                log_info("Not generating service for XDG autostart %s, should be skipped by generator", service->name);
+                return 0;
+        }
+
+        /* Nothing to do if type is not Application. */
+        if (!streq_ptr(service->type, "Application")) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (!service->exec_string) {
+                log_info("Not generating service for XDG autostart %s, it is has no Exec= line", service->name);
+                return 0;
+        }
+
+        /*
+         * The TryExec key cannot be checked properly from the systemd unit,
+         * it is trivial to check using find_binary though.
+         */
+        if (service->try_exec) {
+                r = find_binary(service->try_exec, NULL);
+                if (r < 0) {
+                        log_info("Not generating service for XDG autostart %s, TryExec= binary %s does not exist or not executable",
+                                 service->name, service->try_exec);
+                        return 0;
+                }
+        }
+
+        r = xdg_autostart_format_exec_start(service->exec_string, &exec_start);
+        if (r < 0) {
+                log_info("Not generating service for XDG autostart %s, Exec= line is invalid",
+                         service->name);
```

Comment:
as above: bump to LOG_WARNING if something is invalid but we ignore it. and include error code in output.